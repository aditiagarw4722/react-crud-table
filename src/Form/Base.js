// @flow
import React from 'react';
import { Formik, Field } from 'formik';

import { Form } from './wrappers';
import Button from '../Button';

import { DEFAULT_VALIDATE, generalValidationError } from './helpers';

type FieldData = {
  name: string,
  label: string,
  placeholder?: string,
  render?: Function | string,
  type?: string,
  readOnly?: boolean,
};

type FormBaseProps = {
  touched: Object,
  errors: Object,
  data: {
    message?: string,
    generalErrorMessage?: string,
    fields: Array<FieldData>,
    submitButtonProps: Object,
    submitText: string,
  },
  error: ?string,
};

const FormBase = ({
  data,
  errors,
  touched,
  error,
}: FormBaseProps): React$Element<any> => (
  <Form>
    {data.message && <Form.Message>{data.message}</Form.Message>}
    {error && <Form.ErrorMessage>{error}</Form.ErrorMessage>}
    {generalValidationError({
      touched,
      errors,
      message: data.generalErrorMessage,
    })}
    {data.fields.map((field) => (
      <Form.FieldContainer key={field.name}>
        <Form.Label htmlFor={field.name}>{field.label}</Form.Label>
        <Field
          name={field.name}
          placeholder={field.placeholder}
          render={field.render}
          type={field.type}
          readOnly={field.readOnly}
        />
        {errors[field.name] && touched[field.name] && (
          <Form.FieldError>{errors[field.name]}</Form.FieldError>
        )}
      </Form.FieldContainer>
    ))}
    {generalValidationError({
      touched,
      errors,
      message: data.generalErrorMessage,
    })}
    <Button type="submit" modifiers="positive" {...data.submitButtonProps}>
      {data.submitText}
    </Button>
  </Form>
);

export default FormBase;
