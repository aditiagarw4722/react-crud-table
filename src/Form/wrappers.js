import { Form as FormikForm } from 'formik';
import bcc from 'bem-react-component-creator';

const { block, element } = bcc('crud-modal-form');

export const Form = block(FormikForm);

Form.displayName = 'FormikForm';

Form.Label = element('label', 'label');
Form.Label.displayName = 'Label';

Form.Message = element('div', 'message');
Form.Message.displayName = 'Message';

Form.ErrorMessage = element('div', 'error-message');
Form.ErrorMessage.displayName = 'ErrorMessage';

Form.FieldError = element('div', 'field-error');
Form.FieldError.displayName = 'FieldError';

Form.FieldContainer = element('div', 'field-container');
Form.FieldContainer.displayName = 'FieldContainer';

export default {
  Form,
};
