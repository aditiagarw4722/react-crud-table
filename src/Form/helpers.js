// @flow

import React from 'react';
import { queryValue } from '../CRUDTable/helpers';

import { Form } from './wrappers';

export const DEFAULT_VALIDATE = (): Object => ({});

type Payload = {
  touched: Object,
  errors: Object,
  message: ?string,
};

export const generalValidationError = (
  payload: Payload
): boolean | React$Element<any> => {
  const showErrorMessage = Object.keys(payload.errors).reduce(
    (acc, key) => acc || payload.touched[key],
    false
  );
  return (
    showErrorMessage && (
      <Form.ErrorMessage>
        {queryValue(payload, 'message', 'There are some errors')}
      </Form.ErrorMessage>
    )
  );
};
