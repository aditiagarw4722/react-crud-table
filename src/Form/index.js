import React from 'react';
import PropTypes from 'prop-types';
import { Formik, Field } from 'formik';
import { Form } from './wrappers';
import Button from '../Button';

const DEFAULT_VALIDATE = () => ({});

const generalValidationError = ({ touched, errors, message }) => {
  const fieldsTouched = touched || {};
  const showErrorMessage = Object.keys(errors || {}).reduce(
    (acc, key) => acc || fieldsTouched[key],
    false
  );
  return (
    showErrorMessage && (
      <Form.ErrorMessage>
        {message || 'There are some errors'}
      </Form.ErrorMessage>
    )
  );
};

const BasicForm = ({ data, onSubmit, initialValues }) => (
  <div>
    <Formik
      enableReinitialize
      initialValues={initialValues}
      validate={data.validate || DEFAULT_VALIDATE}
      onSubmit={onSubmit}
      render={({ errors, touched, error }) => (
        <Form>
          {data.message && <Form.Message>{data.message}</Form.Message>}
          {error && <Form.ErrorMessage>{error}</Form.ErrorMessage>}
          {generalValidationError({
            touched,
            errors,
            message: data.generalErrorMessage,
          })}
          {data.fields.map((field) => (
            <Form.FieldContainer key={field.name}>
              <Form.Label htmlFor={field.name}>{field.label}</Form.Label>
              <Field
                name={field.name}
                placeholder={field.placeholder}
                render={field.render}
                type={field.type}
                readOnly={field.readOnly}
              />
              {errors[field.name] && touched[field.name] && (
                <Form.FieldError>{errors[field.name]}</Form.FieldError>
              )}
            </Form.FieldContainer>
          ))}
          {generalValidationError({
            touched,
            errors,
            message: data.generalErrorMessage,
          })}
          <Button
            type="submit"
            modifiers="positive"
            {...(data.submitButtonProps || {})}
          >
            {data.submitText}
          </Button>
        </Form>
      )}
    />
  </div>
);

BasicForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  data: PropTypes.instanceOf(Object),
  initialValues: PropTypes.instanceOf(Object),
};

BasicForm.defaultProps = {
  initialValues: {},
  data: {
    message: null,
    fields: [],
    submitText: null,
  },
};

export default BasicForm;
