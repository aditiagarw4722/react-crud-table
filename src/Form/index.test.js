import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { create } from 'react-test-renderer';

import Form from './';

jest.mock('formik', () => ({
  Formik: 'Formik',
}));

jest.mock('./Base', () => ({
  __esModule: true,
  default: 'FormBase',
}));

import { Formik } from 'formik';

const renderer = new ShallowRenderer();

describe('Form', () => {
  let props;

  beforeEach(() => {
    props = {
      data: {
        title: 'Test title',
        fields: [],
      },
      onSubmit: jest.fn(() => Promise.resolve({ success: true })),
      initialValues: { name: 'John' },
    };
  });

  it('renders as expected with the given props', () => {
    const result = renderer.render(<Form {...props} />);
    expect(result).toMatchSnapshot();
  });

  it('notifies on Submit', () => {
    const result = create(<Form {...props} />);

    result.root.findByType(Formik).props.onSubmit();

    expect(props.onSubmit).toHaveBeenCalledTimes(1);
  });

  it('it renders the internal Form without any issue', () => {
    const result = create(<Form {...props} />);

    result.root.findByType(Formik).props.render({});
  });
});
