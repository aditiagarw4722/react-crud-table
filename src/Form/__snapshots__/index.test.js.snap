// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders as expected with the given props 1`] = `
<div>
  <Formik
    enableReinitialize={true}
    initialValues={
      Object {
        "name": "John",
      }
    }
    onSubmit={[MockFunction]}
    render={[Function]}
    validate={[Function]}
  />
</div>
`;
